### PR TITLE
Add basic save state system

### PR DESCRIPTION
**Pull request by GnomedDev**
Allows users (or the mod, for a future practice mode) to save the state of a kart and restore it later, without needing heavy dolphin savestates. This technically saves a lot more than necessary, as I went through trying to find what would stop desyncs, but the headers could be useful for future use.

`KartSaveState` is currently sized at `4560` (`0x11d0`), so it isn't a massive memory hog.

This could be bound to a button, which I may do in this PR or a future PR depending on how long it takes to merge, but is currently bound to the `/store` and `/reload` commands.

### CODE REVIEW

File: configure.py [hunk 1]
```diff
@@ -1378,6 +1378,7 @@
         os.path.join('payload', 'game', 'kart', 'KartMove.cc'),
         os.path.join('payload', 'game', 'kart', 'KartObject.S'),
         os.path.join('payload', 'game', 'kart', 'KartObject.cc'),
+        os.path.join('payload', 'game', 'kart', 'KartSaveState.cc'),
```
Comment (stblr) [inline configure.py:1381]: Ordering

File: payload/egg/core/eggSystem.cc [hunk 2]
```diff
@@ -6,6 +6,8 @@ extern "C" {
 #include <sp/Commands.h>
 #include <sp/IOSDolphin.hh>
 #include <game/ui/SectionManager.hh>
+#include <game/kart/KartSaveState.hh>
+#include <game/kart/KartObjectManager.hh>
```
Comment (stblr) [inline payload/egg/core/eggSystem.cc:10]: Ordering

File: payload/game/kart/KartSettings.hh [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 
 namespace Kart {
 
 struct KartSettings {
-    u8 _00[0x10 - 0x00];
+    u32 isBike;
+    u32 vehicle;
+    u32 character;
+    u16 wheelCount0; // Used for initialising suspension
+    u16 wheelCount1; // Used for intiialising tires
```
Comment (stblr) [inline payload/game/kart/KartSettings.hh:12]: Could just be renamed to `susCount` and `tireCount`.

File: payload/game/kart/KartObjectProxy.hh [hunk 4]
```diff
@@ -10,18 +10,34 @@ class KartMove;
...
 class KartState;
 class VehiclePhysics;
+class KartBody;
+class KartAction;
+class KartSus;
+class KartSettings;
+class Kart5c;
+class KartSub;
+class KartTire;
+class PlayerModel;
```
Comment (stblr) [inline payload/game/kart/KartObjectProxy.hh:20]: Ordering

File: payload/game/ui/page/RacePage.cc [hunk 5]
```diff
@@ -49,9 +50,16 @@ RacePage *RacePage::Instance() {
...
+    Kart::kartSaveState = new Kart::KartSaveState;
     m_lastWatchedPlayerId = m_watchedPlayerId;
 }
 
+void RacePage::onDeinit() {
+    REPLACED(onDeinit)();
+
+    delete Kart::kartSaveState;
+}
+
```
Comment (stblr) [inline payload/game/ui/page/RacePage.cc:62]: Does this really belong here?

File: payload/game/kart/KartSettings.hh [hunk 6]
```diff
@@ -1,14 +1,19 @@
...
     u8 playerId;
     u8 _11[0x3c - 0x11];
 };
+
```
Comment (stblr) [inline payload/game/kart/KartSettings.hh:16]: No need for a newline.

File: payload/game/kart/KartSus.hh [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#include "KartObjectProxy.hh"
+
+namespace Kart {
+
+class KartSaveState;
+
+struct KartSusBase {
+    u8 _80[0x80 - 0x00];
+};
```
Comment (stblr) [inline payload/game/kart/KartSus.hh:9]: Is this an actual thing?

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/game/kart/KartSus.hh (hunk 7 above)
Comment (GnomedDev) [reply payload/game/kart/KartSus.hh:9]: Ah, this was KartPart in ghidra, I misread it.

File: payload/game/ui/page/RacePage.cc (hunk 5 above)
Comment (GnomedDev) [reply payload/game/ui/page/RacePage.cc:62]: It was the first place I could think of that I could use the heap and is loaded only during the race, where else do you suggest?

Comment (stblr) [reply payload/game/ui/page/RacePage.cc]: `KartObjectManager::CreateInstance`, the only issue is that we are already replacing this function from C code.

File: payload/sp/SaveStateManager.cc [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+#include <tuple>
+
+namespace SP {
+
+SaveStateManager *SaveStateManager::s_instance = nullptr;
+
+SaveStateManager::SaveStateManager() {
+    m_framesHeld = 0;
+    m_hasSaved = false;
+    m_kartSaveState = new Kart::KartSaveState;
```
Comment (stblr) [inline payload/sp/SaveStateManager.cc:17]: This leaks memory.

Comment (stblr) [reply payload/sp/SaveStateManager.cc:17]: By the way, do we even need a separate allocation?

Comment (GnomedDev) [reply payload/sp/SaveStateManager.cc:17]: Thinking about it, probably not, I implemented this way to lazy allocate the save state but that doesn't seem possible or useful (see dev chat last night).

**Review by stblr — APPROVED** [review]

**Pull request merged**